{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "'#number' in each cell represent the sequence to run the cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1\n",
    "import os\n",
    "import pandas as pd\n",
    "import csv\n",
    "from colour import Color\n",
    "import matplotlib\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#edit the path to folder and run at number 2\n",
    "path='.../data-set 3'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#edit the exportpath as per your machine and reun at number 3\n",
    "exportpath='.../out_data/out/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#12\n",
    "#this function takes the path of the folder as input and \n",
    "#returns the two dataframes- variant dataframe and Phenotype dataframe \n",
    "# this function gives call to several other function doing subtasks\n",
    "\n",
    "def starthere(path):\n",
    "    variants_df=pd.DataFrame()\n",
    "    phenotype_df=pd.DataFrame()\n",
    "    \n",
    "    for folder in os.listdir(path):\n",
    "        if folder.startswith('.DS_Store'):\n",
    "            continue\n",
    "        \n",
    "        curr_path=\"\".join((path, \"/\",folder,))\n",
    "        files=os.listdir(curr_path)\n",
    "        if '.DS_Store' in files:\n",
    "            files=files.remove('.DS_Store')\n",
    "        \n",
    "       \n",
    "        varpath=\"\".join((curr_path,\"/\",files[1]))\n",
    "        if varpath.endswith('.txt'):\n",
    "            temp=readvar(varpath)\n",
    "            processed=process(temp)\n",
    "            processed.loc[:,'person_id'] = folder\n",
    "            variants_df=variants_df.append(processed)\n",
    "    \n",
    "      \n",
    "        \n",
    "        phenopath=\"\".join((curr_path,\"/\",files[0]))\n",
    "        \n",
    "        if phenopath.endswith('.txt'):\n",
    "            temp=readpheno(phenopath)\n",
    "            temp['person_id']=''\n",
    "            temp.loc[:,'person_id'] = folder\n",
    "           \n",
    "            phenotype_df=phenotype_df.append(temp)\n",
    "            \n",
    "\n",
    "    \n",
    "    phenotype_df=normalize(pd.DataFrame(phenotype_df))\n",
    "    \n",
    "    return variants_df,phenotype_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-9-04a635a261a0>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  file['genotype']=file[3]+file[4]\n"
     ]
    }
   ],
   "source": [
    "#13\n",
    "variant_database,phenotype_database=starthere(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>genotype</th>\n",
       "      <th>person_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs4345758</td>\n",
       "      <td>1</td>\n",
       "      <td>28663</td>\n",
       "      <td>TT</td>\n",
       "      <td>PY633MJ399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4477212</td>\n",
       "      <td>1</td>\n",
       "      <td>72017</td>\n",
       "      <td>AA</td>\n",
       "      <td>PY633MJ399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1856862</td>\n",
       "      <td>1</td>\n",
       "      <td>110905</td>\n",
       "      <td>AA</td>\n",
       "      <td>PY633MJ399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4096703</td>\n",
       "      <td>1</td>\n",
       "      <td>160593</td>\n",
       "      <td>GG</td>\n",
       "      <td>PY633MJ399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs2994597</td>\n",
       "      <td>1</td>\n",
       "      <td>593369</td>\n",
       "      <td>--</td>\n",
       "      <td>PY633MJ399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677860</th>\n",
       "      <td>rs41534744</td>\n",
       "      <td>26</td>\n",
       "      <td>16129</td>\n",
       "      <td>AA</td>\n",
       "      <td>285536LN7G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677861</th>\n",
       "      <td>rs41419246</td>\n",
       "      <td>26</td>\n",
       "      <td>16145</td>\n",
       "      <td>GG</td>\n",
       "      <td>285536LN7G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677862</th>\n",
       "      <td>rs41466049</td>\n",
       "      <td>26</td>\n",
       "      <td>16162</td>\n",
       "      <td>AA</td>\n",
       "      <td>285536LN7G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677863</th>\n",
       "      <td>rs41355449</td>\n",
       "      <td>26</td>\n",
       "      <td>16327</td>\n",
       "      <td>CC</td>\n",
       "      <td>285536LN7G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677864</th>\n",
       "      <td>rs869031877</td>\n",
       "      <td>26</td>\n",
       "      <td>16391</td>\n",
       "      <td>GG</td>\n",
       "      <td>285536LN7G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9091972 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               rsid chromosome position genotype   person_id\n",
       "1         rs4345758          1    28663       TT  PY633MJ399\n",
       "2         rs4477212          1    72017       AA  PY633MJ399\n",
       "3         rs1856862          1   110905       AA  PY633MJ399\n",
       "4         rs4096703          1   160593       GG  PY633MJ399\n",
       "5         rs2994597          1   593369       --  PY633MJ399\n",
       "...             ...        ...      ...      ...         ...\n",
       "677860   rs41534744         26    16129       AA  285536LN7G\n",
       "677861   rs41419246         26    16145       GG  285536LN7G\n",
       "677862   rs41466049         26    16162       AA  285536LN7G\n",
       "677863   rs41355449         26    16327       CC  285536LN7G\n",
       "677864  rs869031877         26    16391       GG  285536LN7G\n",
       "\n",
       "[9091972 rows x 5 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#14\n",
    "variant_database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Characteristic</th>\n",
       "      <th>Variation</th>\n",
       "      <th>person_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Handedness</td>\n",
       "      <td>Right-handed</td>\n",
       "      <td>PY633MJ399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eye color</td>\n",
       "      <td>brown</td>\n",
       "      <td>PY633MJ399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABO Rh</td>\n",
       "      <td>A-</td>\n",
       "      <td>SEBNF3PGUJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Blood type</td>\n",
       "      <td>A-</td>\n",
       "      <td>SEBNF3PGUJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Interest in Spirituality and Mysticism</td>\n",
       "      <td>None</td>\n",
       "      <td>SEBNF3PGUJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Handedness</td>\n",
       "      <td>Right-handed</td>\n",
       "      <td>285536LN7G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Eye color</td>\n",
       "      <td>brown</td>\n",
       "      <td>285536LN7G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Lisp</td>\n",
       "      <td>No</td>\n",
       "      <td>285536LN7G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Sweat eating spicy food</td>\n",
       "      <td>Sometimes</td>\n",
       "      <td>285536LN7G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Reading habits</td>\n",
       "      <td>Mostly non-fiction</td>\n",
       "      <td>285536LN7G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>376 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Characteristic            Variation   person_id\n",
       "0                               Handedness         Right-handed  PY633MJ399\n",
       "1                                Eye color                brown  PY633MJ399\n",
       "0                                   ABO Rh                   A-  SEBNF3PGUJ\n",
       "1                               Blood type                   A-  SEBNF3PGUJ\n",
       "2   Interest in Spirituality and Mysticism                 None  SEBNF3PGUJ\n",
       "..                                      ...                 ...         ...\n",
       "4                               Handedness         Right-handed  285536LN7G\n",
       "5                                Eye color                brown  285536LN7G\n",
       "6                                     Lisp                   No  285536LN7G\n",
       "7                  Sweat eating spicy food            Sometimes  285536LN7G\n",
       "8                           Reading habits   Mostly non-fiction  285536LN7G\n",
       "\n",
       "[376 rows x 3 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#15\n",
    "phenotype_database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#16\n",
    "export_to_csv(phenotype_database,exportpath,'phenotype_database')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#17\n",
    "export_to_csv(variant_database,exportpath,'variant_database')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#4\n",
    "# input= file: dataframe to export, exportpath, name- name of the csv file you want to create without .csv extension\n",
    "#output= csv file will be created in the file system\n",
    "def export_to_csv(file,exportpath,name):\n",
    "    path=str(exportpath+name+\".csv\")\n",
    "    file.to_csv(path,index = False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#5\n",
    "#input= variant dataset\n",
    "#output= normalized variant data that has cossistent values green-blue and blue-green are entred as one value\n",
    "def normalize(vari):\n",
    "    color=[]\n",
    "    l=[]\n",
    "    seen=[]\n",
    "\n",
    "    for name, hex in matplotlib.colors.cnames.items():\n",
    "        color.append(name)\n",
    "        \n",
    "    for ind,var in enumerate(vari['Variation']):\n",
    "        var=var.lstrip()\n",
    "        var=var.rstrip()\n",
    "        var=var.lower()\n",
    "        if var.startswith(tuple(color)):\n",
    "            l.append([ind,var])\n",
    "           \n",
    "    import re\n",
    "    \n",
    "    for i in l:\n",
    "        ind=i[0]\n",
    "        res = re.split(',|_|-|!| ', i[1])\n",
    "        \n",
    "        for word in res:\n",
    "            if word not in color:\n",
    "                res.remove(word)\n",
    "                \n",
    "        ressort=sorted(res)\n",
    "    \n",
    "        if res not in seen:\n",
    "            seen.append(res)\n",
    "        newst='-'.join(ressort)\n",
    "      \n",
    "        vari.iat[ind,1]=newst\n",
    "\n",
    "    return vari"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#6\n",
    "#input= path of variant.txt file\n",
    "#output= returns the data read from file in file system\n",
    "# this function reads the variant.txt files\n",
    "def readvar(variant):\n",
    "    skip_rows = 0\n",
    "    with open(variant, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('#'):\n",
    "                skip_rows += 1\n",
    "            else:\n",
    "                break\n",
    "    \n",
    "    with open(variant, newline='') as csvfile:\n",
    "        dialect = csv.Sniffer().sniff(csvfile.read(1024))\n",
    "    sepe = dialect.delimiter\n",
    "    if sepe!=',':\n",
    "        sepe='\\t'\n",
    "    file=pd.read_csv(variant,comment='#',skiprows=skip_rows,sep=sepe,header=None,error_bad_lines=False,low_memory=False)\n",
    "    #print('in readvar ',file[:5])\n",
    "    return file\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#7\n",
    "#input= path of Phenotype.txt file\n",
    "#output= returns the data read from file in file system\n",
    "# this function reads the phenitype.txt files\n",
    "def readpheno(pheno):\n",
    "    file= pd.read_csv(pheno,sep='\\t',error_bad_lines=False)\n",
    "    file=pd.DataFrame(file)\n",
    "    return file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#8\n",
    "#input= variant.txt file with 6 columns\n",
    "#output= processed and cleaned file as per the given  and standard guidelines\n",
    "#this function converts genotype(coded with repsect to negative strand ) to genorype that is recorded with repsect to pisitive strand and renames the columns in consistant fashion\n",
    "\n",
    "def process6cols(file):\n",
    "    dic={'A':'T','T':'A','C':'G','G':'C','-':'-'}\n",
    "    if isinstance(file[0][1], str):\n",
    "        file=file[1:]\n",
    "    for index, row in file.iterrows():\n",
    "        code=row[5]\n",
    "        if row[4]=='-':\n",
    "            file.at[index,'genotype'] = dic[code[0]]+dic[code[1]]\n",
    "        else:\n",
    "            file.at[index,'genotype'] = code\n",
    "    \n",
    "    file=file.rename(columns={0: 'rsid', 2: 'chromosome',3:'position'})\n",
    "    file.drop([1,4,5],axis=1,inplace=True)\n",
    "    return pd.DataFrame(file)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#9\n",
    "#input= variant.txt file with 5 columns\n",
    "#output= processed and cleaned file as per the given  and standard guidelines\n",
    "#this function merges two saperate strands in one column-genotype as required, renames the columns in consistant fashion\n",
    "def process5cols(file):\n",
    "    if isinstance(file[0][1], str):\n",
    "        file=file[1:]\n",
    "    file['genotype']=file[3]+file[4]\n",
    "    file=file.rename(columns={0: 'rsid', 1: 'chromosome',2:'position'})\n",
    "    file=file.drop([3,4], axis = 1)\n",
    "    file['genotype']=file['genotype'].replace({'00':'--'},regex=True)\n",
    "    return pd.DataFrame(file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#10\n",
    "#input= variant.txt file with 4 columns\n",
    "#output= processed and cleaned file as per the given  and standard guidelines\n",
    "#renames the columns in consistant fashion, removes the comment \n",
    "def process4cols(file):\n",
    "    #print(file.columns)\n",
    "    if isinstance(file[0][1], str):\n",
    "        file=file[1:]\n",
    "    file=file.rename(columns={0: 'rsid', 1: 'chromosome',2:'position',3:'genotype'})\n",
    "    file['genotype']=file['genotype'].replace({'00':'--'},regex=True)\n",
    "    return pd.DataFrame(file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#11\n",
    "# input= variant file\n",
    "# output= returns the cleaned variant file\n",
    "# this function calls different cleaning function to clean the files in different format\n",
    "# it checks for the number of columns in file, since number of columns is deciding factor in file format in this case\n",
    "def process(file):\n",
    "    cols = len(file.axes[1])\n",
    "    if cols==6:\n",
    "        data6=process6cols(file)\n",
    "\n",
    "        return data6\n",
    "    elif cols==5:\n",
    "        data5=process5cols(file)\n",
    "\n",
    "        return data5\n",
    "    elif cols==4:\n",
    "        data4=process4cols(file)\n",
    "        return data4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#18\n",
    "# creating database with unique combination of characteristics and variants \n",
    "char_var=phenotype_database[['Characteristic ','Variation']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#19\n",
    "# exporting database with unique combination of characteristics and variants \n",
    "export_to_csv(char_var,exportpath,'char_var')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#20\n",
    "# creating database with unique combination of personid and characteristics \n",
    "person_char=phenotype_database[['person_id','Characteristic ']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "#21\n",
    "# exporting database with unique combination of personid and characteristics \n",
    "export_to_csv(person_char,exportpath,'person_char')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
